main/autogen/autogen.cc

```diff
@@ -674,4 +676,331 @@ void ParsedFile::classlist(core::Context ctx, vector<string> &out) {
+        root.addDef(ctx, alConfig, move(namedDef));
+    }
+}
+
+void DefTree::addDef(core::Context ctx, const AutoloaderConfig &alCfg, NamedDefinition ndef) {
+    if (!alCfg.include(ndef)) {
+        return;
+    }
+    auto *node = this;
+    for (const auto &part : ndef.nameParts) {
+        auto &child = node->children[part];
+        if (!child) {
+            child = make_unique<DefTree>();
+            child->nameParts = node->nameParts; // NOTE: this could be tracked recursively with push/pop
+            child->nameParts.emplace_back(part);
+        }
+        node = &*child; // TODO There must be a better way to do this
+    }
+    if (ndef.def.defines_behavior) {
+        node->namedDefs.emplace_back(move(ndef));
+    } else {
+        node->nonBehaviorDefs.emplace_back(move(ndef));
+    }
+}
+
+bool AutoloaderConfig::include(const NamedDefinition &nd) const {
+    return !nd.nameParts.empty() && topLevelNamespaceRefs.find(nd.nameParts[0]) != topLevelNamespaceRefs.end();
+}
+
+bool AutoloaderConfig::includePath(string_view path) const {
+    return absl::EndsWith(path, ".rb") &&
+           !sorbet::FileOps::isFileIgnored("", fmt::format("/{}", path), absoluteIgnorePatterns,
+                                           relativeIgnorePatterns);
+}
+
+bool AutoloaderConfig::includeRequire(core::NameRef req) const {
+    return excludedRequireRefs.find(req) == excludedRequireRefs.end();
+}
+
+AutoloaderConfig AutoloaderConfig::enterConfig(core::GlobalState &gs, const realmain::options::AutoloaderConfig &cfg) {
+    AutoloaderConfig out;
+    out.rootDir = cfg.rootDir;
+    out.preamble = cfg.preamble;
+    for (auto &str : cfg.modules) {
+        out.topLevelNamespaceRefs.emplace(gs.enterNameConstant(str));
+    }
+    for (auto &str : cfg.requireExcludes) {
+        out.excludedRequireRefs.emplace(gs.enterNameUTF8(str));
+    }
+    for (auto &nameParts : cfg.sameFileModules) {
+        vector<core::NameRef> &refs = out.sameFileModuleNames.emplace_back();
+        for (auto &name : nameParts) {
+            refs.emplace_back(gs.enterNameConstant(name));
+        }
+    }
+    out.absoluteIgnorePatterns = cfg.absoluteIgnorePatterns;
+    out.relativeIgnorePatterns = cfg.relativeIgnorePatterns;
+    return out;
+}
+
+void DefTree::prettyPrint(core::Context ctx, int level) {
+    auto fileRefToString = [&](const NamedDefinition &nd) -> string_view { return nd.fileRef.data(ctx).path(); };
+    fmt::print("{} [{}]\n", name().show(ctx), fmt::map_join(namedDefs, ", ", fileRefToString));
+    for (auto &[name, tree] : children) {
+        for (int i = 0; i < level; ++i) {
+            fmt::print("  ");
+        }
+        tree->prettyPrint(ctx, level + 1);
+    }
+}
+
+string DefTree::fullName(core::Context ctx) const {
+    return fmt::format("{}",
+                       fmt::map_join(nameParts, "::", [&](core::NameRef nr) -> string_view { return nr.show(ctx); }));
+}
+
+string join(string path, string file) {
+    if (file.empty()) {
```

Comment:
while is file ever empty?

Reply:
Because the root lives at the same directory as the top-level modules and has not path.